hbp_validation_framework/utils.py

```diff
@@ -70,11 +82,295 @@ def _make_js_file(data):
+    Can be run on node(s) having no access to external URLs (i.e. without internet access)
+
+    Returns
+    -------
+    path
+        The absolute path of the generated test result file
+
+    Examples
+    --------
+    >>> test_result_file = utils.run_test_offline(model=model, test_config_file=test_config_file)
+    """
+
+    if not os.path.isfile(test_config_file) :
+        raise Exception("'test_config_file' should direct to file describing the test configuration.")
+
+    # Load the test info from config file
+    with open(test_config_file) as file:
+        test_info = json.load(file)
+
+    # Identify test class path
+    path_parts = test_info["test_instance_path"].split(".")
+    cls_name = path_parts[-1]
+    module_name = ".".join(path_parts[:-1])
+    test_module = import_module(module_name)
+    test_cls = getattr(test_module, cls_name)
+
+    # Read observation data required by test
+    with open(test_info["test_observation_file"], 'r') as file:
+        observation_data = file.read()
+    content_type = mimetypes.guess_type(test_info["test_observation_file"])[0]
+    if content_type == "application/json":
+        observation_data = json.loads(observation_data)
+
+    # Create the :class:`sciunit.Test` instance
+    params = test_info["params"]
+    test = test_cls(observation=observation_data, **params)
+    test.uuid = test_info["test_instance_id"]
+
+    print("----------------------------------------------")
+    print("Test name: ", test.name)
+    print("Test type: ", type(test))
+    print("----------------------------------------------")
+
+    # Check the model
+    if not isinstance(model, sciunit.Model):
+        raise TypeError("`model` is not a sciunit Model!")
+    print("----------------------------------------------")
+    print("Model name: ", model.name)
+    print("Model type: ", type(model))
+    print("----------------------------------------------")
+
+    # Run the test
+    t_start = datetime.utcnow()
+    score = test.judge(model, deep_error=True)
+    t_end = datetime.utcnow()
+
+    print("----------------------------------------------")
+    print("Score: ", score.score)
+    if "figures" in score.related_data:
+        print("Output files: ")
+        for item in score.related_data["figures"]:
+            print(item)
+    print("----------------------------------------------")
+
+    score.runtime = str(int(math.ceil((t_end-t_start).total_seconds()))) + " s"
+    score.exec_timestamp = t_end
+    # score.exec_platform = str(self._get_platform())
+
+    # Save result info to file
+    test_result_file = os.path.join(test_info["base_folder"], "test_result.pkl")
+    with open(test_result_file, 'wb') as file:
+        pickle.dump(score, file)
+    return test_result_file
+
+def upload_test_result(hbp_username="", hbp_password=None, environment="production", test_result_file="", storage_collab_id="", register_result=True, client_obj=None):
+    """Register the result on the validation framework
```

Comment:
Register the result with the Validation Service

Reply:
Done